### PR TITLE
Pipca task

### PR DESCRIPTION
**Pull request by ThomasBnon**
# Main Changes

I've designed two distinct tasks: `pipca_offline` and `pipca_online` to handle Principal Component Analysis (PCA) using the Pipca model in tasks.py.
- For `pipca_online`, the task is streamlined. If a pipca_model.h5 file exists, the script utilizes the prior values of U, S, mu, and var. This promotes efficiency by reusing existing model parameters.
- On the other hand, `pipca_offline` employs a for loop across various runs. In each iteration, U, S, mu, and var are initialized with the values from the preceding run. This iterative approach enables incremental refinement of the PCA model. This tasks allows us to follow the evolution of the compression loss, the time of execution, and the difference in norms between matrixes U, S, mu and var.
Capture
As we don't know which sample is used for each run, important jumps could help us identify a sample change between runs.
- For these tasks, I coded a new function in PiPCA class called `run_model_full` that takes into account previous variables U, S, mu and var.

## Other Changes

- To enhance control, a 'tl' (time limit) command has been implemented in elog_submit, allowing users to set a specific time constraint for the calculation.
- Additionally, I've implemented a function called `compute_compression_loss` to calculate the compression loss using Pipca across all images in pipca_visuals.py. This function ensures a comprehensive assessment of the model's performance.
- Recognizing the potential computational intensity of the task, I've created another function: `compute_compression_loss_random_images` has the capability to operate on a random sample of images, significantly reducing computation time.

## Elog Commands 

To run the job, the following command can be used:
`elog_submit.sh -c ../yamls/mfxp23120.yaml -t pipca_online -n 1 -tl 20:00:00 -a lcls:mfxp23120`
or
`elog_submit.sh -c ../yamls/mfxp23120.yaml -t pipca_offline -n 1 -tl 20:00:00 -a lcls:mfxp23120`

The yaml file is located in `/sdf/data/lcls/ds/mfx/mfxp23120/scratch/thomasbn/yamls`

### CODE REVIEW

File: btx/misc/pipca_visuals.py [hunk 1]
```diff
@@ -277,4 +277,125 @@ def construct_heatmap_data(img, max_pixels):
...
+    # Calculate the average of the norms
+    average_norm = np.mean(image_norms)
+
+    return average_norm
+
+import random
+
+def compute_compression_loss_random_images(filename, num_components, num_images=10):
+    """
+    Compute average Frobenius norms for a random subset of images.
```
Comment (fredericpoitevin) [inline btx/misc/pipca_visuals.py:349]: see comment above.

File: btx/processing/pipca.py [hunk 2]
```diff
@@ -175,6 +175,76 @@ def run_model(self):
                 f.create_dataset('V', data=V)
                 print(f'Model saved to {self.filename}')
 
+    def run_model_full(self, previous_U, previous_S, previous_mu, previous_var):
```
Comment (fredericpoitevin) [inline btx/processing/pipca.py:178]: Likewise, there is a very high level of redundancy with `run_model`. Consider merging them into one.

File: btx/processing/pipca.py [hunk 3]
```diff
@@ -304,6 +374,8 @@ def update_model(self, X):
                 self.task_durations, "center data and compute augment vector"
             ):
                 X_centered = X - np.tile(mu_m, m)
+
+
```
Comment (fredericpoitevin) [inline btx/processing/pipca.py:378]: Are those empty lines here intentionally?

File: scripts/elog_submit.sh [hunk 4]
```diff
@@ -25,6 +25,8 @@ $(basename "$0"):
       Run Number
     -t|--task
       Task name
+    -tl|--time_limit
+      change time limit
```
Comment (fredericpoitevin) [inline scripts/elog_submit.sh:29]: nice addition!

File: scripts/main.py [hunk 5]
```diff
@@ -66,3 +66,5 @@ def main():
 
     print(status_message)
     exit(retval)
+
+#shdhfhiqskudhfsudfhghsdjf
```
Comment (fredericpoitevin) [inline scripts/main.py:70]: everything ok? :D

File: scripts/tasks.py [hunk 6]
```diff
@@ -509,3 +510,242 @@ def timetool_correct(config):
...
+        os.remove(filename_with_tag)
+
+    # Initialize U, S, mu and var matrices
+    if os.path.exists(filename_with_tag):
+        with h5py.File(filename_with_tag, 'r') as f:
+            previous_U = np.asarray(f.get('U'))
+            previous_S = np.asarray(f.get('S'))
+            previous_mu_tot = np.asarray(f.get('mu'))
+            previous_var_tot = np.asarray(f.get('total_variance'))
+    else:
```
Comment (fredericpoitevin) [inline scripts/tasks.py:546]: you can put the overwrite test under the existence test.

File: scripts/tasks.py [hunk 7]
```diff
@@ -509,3 +510,242 @@ def timetool_correct(config):
...
+        execution_times.append(execution_time)  # Append the execution time to the list
+
+        U = pipca.gather_U()
+        S = pipca.S
+        mu_tot = pipca.gather_mu()
+        var_tot = pipca.gather_var()
+
+        # Compute norms change
+        diff_U, diff_S, diff_mu, diff_var = (U - previous_U, S - previous_S, mu_tot - previous_mu_tot, var_tot - previous_var_tot) if previous_U is not None else (U, S, mu_tot, var_tot)
+        norm_diff_U_list.append(np.linalg.norm(diff_U)); norm_diff_S_list.append(np.linalg.norm(diff_S)); norm_diff_mu_list.append(np.linalg.norm(diff_mu)); norm_diff_var_list.append(np.linalg.norm(diff_var))
```
Comment (fredericpoitevin) [inline scripts/tasks.py:607]: I think it's OK to "decompactify" these two lines, to increase readability. Maybe consider creating a function for this?

File: scripts/tasks.py [hunk 8]
```diff
@@ -509,3 +510,242 @@ def timetool_correct(config):
...
+                f['norm_diff_mu_list'][-1] = np.linalg.norm(diff_mu)
+
+            # Check if 'norm_diff_var_list' dataset exists, create if not
+            if 'norm_diff_var_list' not in f:
+                f.create_dataset('norm_diff_var_list', data=np.array(norm_diff_var_list))
+            else:
+                # Append norm_diff_var_list to the existing dataset
+                existing_norm_diff_var_list = f['norm_diff_var_list'][:]
+                f['norm_diff_var_list'].resize((len(existing_norm_diff_var_list) + 1,))
+                f['norm_diff_var_list'][-1] = np.linalg.norm(diff_var)
```
Comment (fredericpoitevin) [inline scripts/tasks.py:672]: Maybe consider creating a function to do all this?

File: scripts/tasks.py [hunk 9]
```diff
@@ -509,3 +510,242 @@ def timetool_correct(config):
...
+
+
+def pipca_online(config):
```
Comment (fredericpoitevin) [inline scripts/tasks.py:675]: Redundancy is high between pipca_online and pipca_offline. Let's try to fix this.

**Review by fredericpoitevin — CHANGES_REQUESTED** [review]
It's great, thanks Thomas! Before we merge, let's address the redundancy issues. We can look into it together. Thanks!

File: btx/misc/pipca_visuals.py [hunk 10]
```diff
@@ -277,4 +277,125 @@ def construct_heatmap_data(img, max_pixels):
     # Create hm_data array for heatmap
     hm_data = np.stack((rows, cols, binned_img.flatten())).T
 
-    return hm_data
\ No newline at end of file
+    return hm_data
+
+def compute_compression_loss(filename, num_components):
+    """
+    Compute average frobenius norms
```
[the remark this answers is not in the document]
Comment (ThomasBnon) [reply btx/misc/pipca_visuals.py:284]: I did the modifications

File: btx/misc/pipca_visuals.py [hunk 11]
```diff
@@ -277,4 +277,125 @@ def construct_heatmap_data(img, max_pixels):
...
+    psi.counter = start_img
+
+    # Create PC dictionary and widgets
+    PCs = {f'PC{i}' : v for i, v in enumerate(loadings, start=1)}
+
+    # Create a list to store the Frobenius norms of the differences between images and their reconstructed counterparts
+    image_norms = []
+
+    # Compute the projection matrix outside the loop
+    projection_matrix = U[:, :q] @ np.diag(S[:q])
```
[the remark this answers is not in the document]
Comment (ThomasBnon) [reply btx/misc/pipca_visuals.py:318]: I added:
```
    model_rank = S.shape[0]
    if num_components > model_rank:
        raise ValueError("Error: num_components cannot be greater than the maximum model rank.")
```
As the maximum number of components corresponds to the number of values on S diagonal, the num_components specified must be lower than the number of values on S diagonal.

I also renamed q to num_components in both functions.

File: btx/misc/pipca_visuals.py [hunk 12]
```diff
@@ -277,4 +277,125 @@ def construct_heatmap_data(img, max_pixels):
...
+    for img_source in range(len(PCs['PC1'])):
+        counter = psi.counter
+        psi.counter = start_img + img_source
+        img = psi.get_images(1)
+        psi.counter = counter
+        img = img.squeeze()
+
+        p, x, y = psi.det.shape()
+        pixel_index_map = retrieve_pixel_index_map(psi.det.geometry(psi.run))
+        q = num_components
```
[the remark this answers is not in the document]
Comment (ThomasBnon) [reply btx/misc/pipca_visuals.py:330]: Yes! I did the modifications.

File: scripts/tasks.py (hunk 7 above)
Comment (ThomasBnon) [reply scripts/tasks.py:607]: Ok! I did this instead:

```
# Compute norms change
        def compute_norm_difference(current, previous=None):
            if previous is not None:
                diff = current - previous
            else:
                diff = current
            return diff

        # Compute norm differences for each list
        diff_U = compute_norm_difference(U, previous_U)
        diff_S = compute_norm_difference(S, previous_S)
        diff_mu = compute_norm_difference(mu_tot, previous_mu_tot)
        diff_var = compute_norm_difference(var_tot, previous_var_tot)

        # Append results to respective lists
        norm_diff_U_list.append(np.linalg.norm(diff_U))
        norm_diff_S_list.append(np.linalg.norm(diff_S))
        norm_diff_mu_list.append(np.linalg.norm(diff_mu))
        norm_diff_var_list.append(np.linalg.norm(diff_var))

        # Update U and S for the next run
        previous_U = U
        previous_S = S
        previous_mu_tot = mu_tot
        previous_var_tot = var_tot
```

File: scripts/tasks.py (hunk 8 above)
Comment (ThomasBnon) [reply scripts/tasks.py:672]: Done!

File: btx/misc/pipca_visuals.py [hunk 13]
```diff
@@ -277,4 +277,125 @@ def construct_heatmap_data(img, max_pixels):
...
+        reconstructed_img = reconstructed_img.reshape((p, x, y))
+        reconstructed_img = assemble_image_stack_batch(reconstructed_img, pixel_index_map)
+
+        # Compute the Frobenius norm of the difference between the original image and the reconstructed image
+        difference = np.abs(img - reconstructed_img)
+        norm = np.linalg.norm(difference, 'fro')
+        image_norms.append(norm)
+
+    # Calculate the average of the norms
+    average_norm = np.mean(image_norms)
```
[the remark this answers is not in the document]
Comment (ThomasBnon) [reply btx/misc/pipca_visuals.py:341]: Yes, I will add these features

File: btx/processing/pipca.py (hunk 2 above)
Comment (ThomasBnon) [reply btx/processing/pipca.py:178]: I merged the two functions:

```
def run_model_full(self, previous_U = None, previous_S = None, previous_mu = None, previous_var = None):
        """
        Perform iPCA on run subject to initialization parameters.
        """
        batch_size = self.batch_size
        num_images = self.num_images

        # initialize and prime model, if specified
        if self.priming:
            img_batch = self.get_formatted_images(
                self.num_components, 0, self.num_features
            )
            self.prime_model(img_batch)

        elif previous_U is not None:
            self.U = previous_U
            self.S = previous_S
            self.mu = previous_mu
            self.total_variance = previous_var

        else:
            self.U = np.zeros((self.split_counts[self.rank], self.num_components))
            self.S = np.ones(self.num_components)
            self.mu = np.zeros((self.split_counts[self.rank], 1))
            self.total_variance = np.zeros((self.split_counts[self.rank], 1))

        # divide remaining number of images into batches
        # will become redundant in a streaming setting, need to change
        rem_imgs = num_images - self.num_incorporated_images
        batch_sizes = np.array(
            [batch_size] * np.floor(rem_imgs / batch_size).astype(int)
            + ([rem_imgs % batch_size] if rem_imgs % batch_size else [])
        )

        # define batch indices based on batch sizes
        self.batch_indices = self.distribute_images_over_batches(batch_sizes)
        self.batch_number = 0

        # update model with remaining batches
        for batch_size in batch_sizes:
            self.fetch_and_update_model(batch_size)
            
        self.comm.Barrier()
        
        U = self.gather_U()
        S = self.S
        V = self.V
        
        if self.rank == 0:  
            print("Model complete")
        
            # save model to an hdf5 file
            with h5py.File(self.filename, 'w') as f:
                f.create_dataset('exp', data=self.psi.exp)
                f.create_dataset('run', data=self.psi.run)
                f.create_dataset('det_type', data=self.psi.det_type)
                f.create_dataset('start_offset', data=self.start_offset)
                f.create_dataset('loadings', data=self.pc_data)
                f.create_dataset('U', data=U)
                f.create_dataset('S', data=S)
                f.create_dataset('V', data=V)
                f.create_dataset('mu', data=self.mu)
                f.create_dataset('total_variance', data=self.total_variance)
                print(f'Model saved to {self.filename}')
```

File: btx/processing/pipca.py (hunk 3 above)
Comment (ThomasBnon) [reply btx/processing/pipca.py:378]: No, I removed them

File: scripts/tasks.py (hunk 9 above)
Comment (ThomasBnon) [reply scripts/tasks.py:675]: I changed the function using this:

```
offline = task.offline

    if offline:
        start_run = run
        num_run = task.max_run
        
    else:
        start_run = run
        num_run = 1
```
In the config file we specify if we want to use the "mode" offline or online.

**Review by fredericpoitevin — APPROVED** [review]
Great work!

**Pull request merged**